### PR TITLE
PauliEvolution gate

### LINKED ISSUE #6863: Evolution gates
### What is the expected behavior?

Implement an evolution gate that can get synthesized into Pauli-evolution gates (or for a simulator backend just with direct matrix exponentiation) such that the compiler can perform optimizations on blocks of evolved Pauli operators before further decomposing into basis gates. 

Here's a preliminary design doc: https://gist.github.com/Cryoris/abc9753fd7dfb0cc3b3a83aa58a182e3

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Implement an evolution gate for the evolution of sums (or sums of sums) of Paulis. The evolution gate is an opaque block containing all evolution information so the compiler can select do the synthesis and do optimizations.

Closes #6863.

### Details and comments

In a nutshell:
* The `PauliEvolutionGate` contains the evolution info (Operator and evolution time) and can be equipped with an `EvolutionSynthesis` to specify how to unroll. Eventually the compiler should select that.
* The `EvolutionSynthesis` determines how to synthesize the evolution and for now includes
    * product formulas: `LieTrotter` and `SuzukiTrotter`
    * exact, matrix-based synthesis: `MatrixExponentiation`

Optimizations we can do with this:
* Pulse-efficient decomp of two-qubit rotations (https://arxiv.org/pdf/2105.01063.pdf)
* Exploit the fact that evolutions of commuting Pauli's commute

Design doc: https://gist.github.com/Cryoris/abc9753fd7dfb0cc3b3a83aa58a182e3

### CODE REVIEW

File: qiskit/circuit/library/evolution/evolution_synthesis.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    """
+
+    @abstractmethod
+    def synthesize(
+        self, operators: Union[SparsePauliOp, List[SparsePauliOp]], time: float
+    ) -> QuantumCircuit:
+        """Synthesize the evolution of a (list of) operator(s) to a circuit.
+
+        If a single operator, the summands are assumed not to commute.
+        If a list, the lust elements are *not* assumed to commute pairwisely, however summands in
```
Comment (woodsp-ibm) [inline qiskit/circuit/library/evolution/evolution_synthesis.py:34]:
```suggestion
        If a list, the list elements are *not* assumed to commute pairwisely, however summands in
```
I imagine this is what was meant - though `its elements are...` would work too

**Comment by woodsp-ibm** [comment]
Needs a reno release note

File: qiskit/circuit/library/evolution/lie_trotter.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        atomic_evolution: Optional[Callable[[SparsePauliOp, float], QuantumCircuit]] = None,
+        insert_barriers: bool = False,
+        cx_structure: str = "chain",
+    ) -> None:
+        """
+        Args:
+            order: The order of the product formula.
+            reps: The number of time steps.
+            atomic_evolution: A function to construct the circuit for the evolution of single operators.
+                Per default, `PauliEvolutionGate` will be used.
```
Comment (woodsp-ibm) [inline qiskit/circuit/library/evolution/lie_trotter.py:38]: Args do not match to init

File: qiskit/circuit/library/evolution/evolution_synthesis.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+from typing import List, Union
+from abc import ABC, abstractmethod
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.quantum_info import SparsePauliOp
+
+
+class EvolutionSynthesis(ABC):
+    """Interface for evolution synthesis algorithms.
+
+    TODO do we need this or do we prefer plain functions?
```
Comment (woodsp-ibm) [inline qiskit/circuit/library/evolution/evolution_synthesis.py:24]: I guess this is TBD

File: qiskit/circuit/library/evolution/pauli_evolution.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    def __init__(
+        self,
+        pauli: Pauli,
+        time: Union[float, ParameterExpression] = 1.0,
+        cx_structure: str = "chain",
+        label: Optional[str] = None,
+    ) -> None:
+        """
+        Args:
+            operator: The Pauli to evolve.
```
Comment (woodsp-ibm) [inline qiskit/circuit/library/evolution/pauli_evolution.py:34]:
```suggestion
            pauli: The Pauli to evolve.
```
Its called `pauli` above

File: qiskit/circuit/library/evolution/suzuki_trotter.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        order: int = 2,
+        atomic_evolution: Optional[Callable[[SparsePauliOp, float], QuantumCircuit]] = None,
+        insert_barriers: bool = False,
+        cx_structure: str = "chain",
+    ) -> None:
+        """
+        Args:
+            order: The order of the product formula.
+            reps: The number of time steps.
+            atomic_evolution: A function to construct the circuit for the evolution of single operators.
```
Comment (woodsp-ibm) [inline qiskit/circuit/library/evolution/suzuki_trotter.py:38]: Missing some args - maybe `order` and `reps` would be preferable in same order as init ie reps then order...

**Review by ajavadia — APPROVED** [review]

**Pull request merged**